Misplaced "type of this value must be known ..." error

In a transformation sequence, I got a type inference error on the first function call, while the error was actually later (and below) in the chain. This was very confusing, in particular since the offending line was not shown in the error message. This could be a duplicate of #36495, or a more general issue not related to type inference.

As it should, rustc 1.11.0 on OS X fails to build the following code:

```
fn fail(v: Vec<u32>) -> Option<Box<Vec<u32>>> {
    v.iter()
        .map(|x| Some(*x))
        .collect()
        .map(Box::new)
}

fn main() {
    fail(vec![1, 2, 3]);
}
```

The error message is:

```
/tmp/err.rs:2:5: 5:23 error: the type of this value must be known in this context
/tmp/err.rs:2     v.iter()
                  ^
error: aborting due to previous error
```

On nightly the error report is nicer, but no more helpful. The fix, which is obvious once one decides to ignore the line indication from rustc, is to add a type annotation to `collect()`:

```
fn fail(v: Vec<u32>) -> Option<Box<Vec<u32>>> {
    v.iter()
        .map(|x| Some(*x))
        .collect::<Option<_>>()
        .map(Box::new)
}

fn main() {
    fail(vec![1, 2, 3]);
}
```

After poking around a bit, I suspect that this may be a more general issue related to how rustc determines the offending line number. For instance, the error message from the compiler has the same problem, while the error is "mismatched types":

```
fn fail(v: Vec<u32>) -> Option<Vec<u32>> {
    v.iter()
        .map(|x| Some(*x))
}

fn main() {
    fail(vec![1, 2, 3]);
}
```

```
/tmp/err.rs:2:5: 3:27 error: mismatched types [E0308]
/tmp/err.rs:2     v.iter()
                  ^

[...]
```

While this is the kind of thing that only bites you once, it can utterly confuse a beginner. It is particularly perplexing in more complex, multi-line expressions. One could of course argue that the error location is _correct_ as it points to the flawed expression, but that does not really help. Perhaps indicating where in the expression the error lies would be better, e.g (using the new and much better error message style from nightly)?

```
error: the type of this value must be known in this context
 --> /tmp/err.rs:2:5
  |
2 |     v.iter()
3 |         .map(|x| Some(*x))
4 |         .collect()
  |         ^

error: aborting due to previous error
```

@Mark-Simulacrum has closed issue #43233, so now this is the additional problem that should be fixed by #36598 (otherwise I'll reopen #43233):

This code:

```
fn sum_and_show(data: &[u32]) -> String {
    data.iter().sum().to_string()
}
fn main() {
    let s: String = sum_and_show(&[1, 2, 3]);
}
```

In rustc 1.20.0-nightly (b2c0707 2017-07-13) gives the error message:

```
error[E0619]: the type of this value must be known in this context
 --> ...\test.rs:2:5
  |
2 |     data.iter().sum().to_string()
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
error[E0619]: the type of this value must be known in this context
 --> ...\test.rs:2:5
  |
2 |     data.iter().sum().to_string()
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

The error message doesn't point to the exact location of the problem. The correct code is:

`data.iter().sum::<u32>().to_string()`

So probably I'd like an error message like:

```
error[E0619]: the type of this value must be known in this context
 --> ...\test.rs:2:5
  |
2 |     data.iter().sum().to_string()
  |                 ^^^^^
```

--- a/src/librustc_typeck/check/mod.rs
+++ b/src/librustc_typeck/check/mod.rs
@@ -2925,7 +2925,7 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         let rcvr = &args[0];
         let rcvr_t = self.check_expr_with_needs(&rcvr, needs);
         // no need to check for bot/err -- callee does that
-        let rcvr_t = self.structurally_resolved_type(expr.span, rcvr_t);
+        let rcvr_t = self.structurally_resolved_type(args[0].span, rcvr_t);
 
         let method = match self.lookup_method(rcvr_t,
                                               segment,

--- a/src/test/ui/span/issue-42234-unknown-receiver-type.rs
+++ b/src/test/ui/span/issue-42234-unknown-receiver-type.rs
@@ -1,0 +1,27 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// When the type of a method call's receiver is unknown, the span should point
+// to the receiver (and not the entire call, as was previously the case before
+// the fix of which this tests).
+
+fn shines_a_beacon_through_the_darkness() {
+    let x: Option<_> = None;
+    x.unwrap().method_that_could_exist_on_some_type();
+    //~^ ERROR 17:5: 17:15: the type of this value must be known in this context
+}
+
+fn courier_to_des_moines_and_points_west(data: &[u32]) -> String {
+    data.iter() //~ ERROR 22:5: 23:20: the type of this value must be known in this context
+        .sum::<_>()
+        .to_string()
+}
+
+fn main() {}

--- a/src/test/ui/span/issue-42234-unknown-receiver-type.stderr
+++ b/src/test/ui/span/issue-42234-unknown-receiver-type.stderr
@@ -1,0 +1,15 @@
+error[E0619]: the type of this value must be known in this context
+  --> $DIR/issue-42234-unknown-receiver-type.rs:17:5
+   |
+17 |     x.unwrap().method_that_could_exist_on_some_type();
+   |     ^^^^^^^^^^
+
+error[E0619]: the type of this value must be known in this context
+  --> $DIR/issue-42234-unknown-receiver-type.rs:22:5
+   |
+22 | /     data.iter() //~ ERROR 22:5: 23:20: the type of this value must be known in this context
+23 | |         .sum::<_>()
+   | |___________________^
+
+error: aborting due to 2 previous errors
+

